Sources/NIOCore/AsyncChannelAdapter.swift

```diff
@@ -0,0 +1,296 @@
+        self.config = config
+        self._lock = Lock()
+        self._state = .bufferingWithoutPendingRead(CircularBuffer(initialCapacity: config.bufferSize))
+    }
+
+    @inlinable
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        // TODO: This should really be done a little differently, with a single big append in the channelReadComplete to reduce
+        // lock contention.
+        var promise: EventLoopPromise<Void>? = nil
+        if self.config.forwardReads {
+            let readPromise = context.eventLoop.makePromise(of: Void.self)
+            readPromise.futureResult.whenSuccess {
+                // This is less than ideal because we pay no attention to read(), but I didn't really
+                // want to manage the complexity of that right now. We should be wary of this though
+                // as it might cause gnarly bugs in the future.
+                context.fireChannelRead(data)
+                context.fireChannelReadComplete()
+            }
+            promise = readPromise
+        }
+        self._receiveNewEvent(.channelRead(self.unwrapInboundIn(data), promise))
+    }
+
+    @inlinable
+    public func channelInactive(context: ChannelHandlerContext) {
+        self._receiveNewEvent(.eof)
+    }
+
+    @inlinable
+    public func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+        if let event = event as? ChannelEvent, event == .inputClosed {
+            self._receiveNewEvent(.eof)
+        }
+    }
+
+    @inlinable
+    public func read(context: ChannelHandlerContext) {
+        let continueReading: Bool = self._lock.withLock {
+            switch self._state {
+            case .bufferingWithPendingRead:
+                return false
+            case .waitingForBuffer:
+                return true
+            case .bufferingWithoutPendingRead(let buffer):
+                if buffer.count >= self.config.bufferSize {
+                    // TODO: Rewrite this to avoid promises.
+                    let promise = context.eventLoop.makePromise(of: Void.self)
+                    promise.futureResult.whenComplete { result in
+                        self._delayedRead(result: result, context: context)
+                    }
+                    self._state = .bufferingWithPendingRead(buffer, promise)
+                    return false
+                } else {
+                    return true
+                }
+            }
+        }
+
+        if continueReading {
+            context.read()
+        }
+    }
+
+    @inlinable
+    func _delayedRead(result: Result<Void, Error>, context: ChannelHandlerContext) {
+        self._lock.withLockVoid {
+            switch self._state {
+            case .bufferingWithPendingRead(let buffer, _):
+                self._state = .bufferingWithoutPendingRead(buffer)
+            case .bufferingWithoutPendingRead, .waitingForBuffer:
+                preconditionFailure()
```

Comment:
I think there is a bug here, if buffer size is small and async reader is fast, we might end up in a situation where `_delayedRead` promise will be completed at an unpredictable time, when state can be `. bufferingWithoutPendingRead` (I'm hitting this constantly with a reader that just accumulates bytes count)

Reply:
Yes agreed.